scipy/signal/tests/test_result_type.py

```diff
@@ -0,0 +1,180 @@
+        (sos, x, sos_zi),
+    ),
+    ("filtfilt", signal.filtfilt, (b, a, x)),
+    (
+        "filtfilt_gust",
+        lambda b, a, x: signal.filtfilt(b, a, x, method="gust"),
+        (b, a, x),
+    ),
+    ("lfiltic(3arg)", signal.lfiltic, (b, a, y)),
+    ("lfiltic(3arg)(len(y)<len(a))", signal.lfiltic, (b, a, y[:1])),
+    # 4-ary
+    (
+        "lfilter[0]",
+        lambda b, a, x, zi: signal.lfilter(b, a, x, zi=zi)[0],
+        (b, a, x, ba_zi),
+    ),
+    (
+        "lfilter[1]",
+        lambda b, a, x, zi: signal.lfilter(b, a, x, zi=zi)[1],
+        (b, a, x, ba_zi),
+    ),
+    ("lfiltic(4arg)", signal.lfiltic, (b, a, y, x)),
+    ("lfiltic(4arg)(len(y)<len(a))", signal.lfiltic, (b, a, y[:1], x)),
+]
+
+# functions without complex support
+no_complex_funcs = [
+    "savgol_filter",
+    "filtfilt_gust",
+]
+
+# functions where all integer args raise NotImplemented
+int_not_impl_funcs = [
+    "lfilter[0]",
+    "lfilter[1]",
+    "filtfilt_gust",
+    "sosfilt[0]",
+    "sosfilt[1]",
+]
+
+test_cases = []
+
+# generate the (f4, f8, c8, c16) -> (f4, f8, c8, c16) test cases
+# per entry in test_base, the test cases are the combinations of (f4,
+# f8, c8, c16) types per argument
+for name, func, arguments in test_base:
+    if name not in no_complex_funcs:
+        typeprod = [(np.float32, np.float64, np.complex64, np.complex128) for i in range(len(arguments))]
+    else:
+        typeprod = [(np.float32, np.float64) for i in range(len(arguments))]
+    implemented = True
+
+    for argtypes in itertools.product(*typeprod):
+        castargs = [arg.astype(argtype)
+                    for arg, argtype in zip(arguments, argtypes)]
+        etype = np.result_type(*castargs)
+        if len(argtypes) > 1:
+            inner = ",".join(typenames[at] for at in argtypes)
+            argtypenames = "(" + inner + ")"
+        else:
+            argtypenames = typenames[argtypes[0]]
+        etypename = typenames[etype]
+
+        testid = "{name}:{argtypenames}->{etypename}".format(
+            name=name, argtypenames=argtypenames, etypename=etypename
+        )
+        test_cases.append(
+            pytest.param(func, castargs, etype, implemented, id=testid)
+        )
+
+
+# generate i8 -> f8 cases
+for name, func, arguments in test_base:
+    implemented = name not in int_not_impl_funcs
+    argtypes = [np.int64 for i in range(len(arguments))]
+    castargs = [arg.astype(argtype)
+                for arg, argtype in zip(arguments, argtypes)]
+    etype = np.float64
+    if len(argtypes) > 1:
+        argtypenames = (
+            "(" + ",".join([typenames[argtype] for argtype in argtypes]) + ")"
+        )
+    else:
+        argtypenames = typenames[argtypes[0]]
+    etypename = typenames[etype]
+    testid = "{name}:{argtypenames}->{etypename}".format(
+        name=name, argtypenames=argtypenames, etypename=etypename
+    )
+    test_cases.append(
+        pytest.param(func, castargs, etype, implemented, id=testid)
+    )
```

Comment:
I agree all of this prep is not worth the code complexity, and the number of tests is not worth the 5 second hit. We actually pay a price for maintainability by making tests too complex. So even though might be a good way to ensure we 100% never break backward compat, it feels overkill.

We can rely on `np.result_type` to do the right thing, so is there some minimal set of combinations we could test (maybe in the 4-8 range) that would expose the bug reported on `master` (and maybe one or two others like it) but that it's fixed on this PR? It would be great if this test took < 100 ms and < 50 lines of code...

Reply:
OK, so more on the lines of a `def test_<func>` per function I actually changed, with just a few test cases exercising the bugs and any behaviour? (The new behaviour is `decimate` converting bool / int / unsigned int to float64.)

Thanks for the review.